vulnerabilities/data_source.py

```diff
@@ -0,0 +1,371 @@
+
+        :param subdir: filter by files in this directory
+        :param recursive: whether to include files in subdirectories
+        :param file_ext: filter files by this extension
+        :return: The first set are added files, the second one modified files
+        """
+        if subdir is None:
+            working_dir = self.config.working_directory
+        else:
+            working_dir = os.path.join(self.config.working_directory, subdir)
+
+        path = Path(working_dir)
+
+        if self.config.last_run_date is None and self.config.cutoff_date is None:
+            if recursive:
+                glob = '**/*'
+            else:
+                glob = '*'
+
+            if file_ext:
+                glob = f'{glob}.{file_ext}'
+
+            return {str(p.relative_to(working_dir)) for p in path.glob(glob) if p.is_file()}, set()
+
+        return self._collect_file_changes(subdir=subdir, recursive=recursive, file_ext=file_ext)
+
+    def added_files(
+            self,
+            subdir: str = None,
+            recursive: bool = False,
+            file_ext: Optional[str] = None
+    ) -> Set[str]:
+
+        return self._collect_files(pygit2.GIT_DELTA_ADDED, subdir, recursive, file_ext)
+
+    def updated_files(
+            self,
+            subdir: str = None,
+            recursive: bool = False,
+            file_ext: str = None
+    ) -> Set[str]:
+
+        if self.config.last_run_date is None and self.config.cutoff_date is None:
+            return set()
+
+        return self._collect_files(pygit2.GIT_DELTA_MODIFIED, subdir, recursive, file_ext)
+
+    def _collect_file_changes(
+            self,
+            subdir: Optional[str],
+            recursive: bool,
+            file_ext: Optional[str],
+    ) -> Tuple[Set[str], Set[str]]:
+
+        last_run = 0 if self.config.last_run_date is None else int(self.config.last_run_date.timestamp())
+        cutoff = 0 if self.config.cutoff_date is None else int(self.config.cutoff_date.timestamp())
+        cutoff = max(last_run, cutoff)
+
+        previous_commit = None
+        added_files, updated_files = set(), set()
+
+        for commit in self._repo.walk(self._repo.head.target, pygit2.GIT_SORT_TIME):
+            commit_time = commit.commit_time + commit.commit_time_offset  # convert to UTC
+
+            if commit_time < cutoff:
+                break
+
+            if previous_commit is None:
+                previous_commit = commit
+                continue
+
+            for d in commit.tree.diff_to_tree(previous_commit.tree).deltas:
+                if not _include_file(d.new_file.path, subdir, recursive, file_ext) or d.is_binary:
+                    continue
+
+                # TODO
+                # Just filtering on the two status values for "added" and "modified" is too simplistic.
+                # This does not cover file renames, copies & deletions.
```

Comment:
This can be solved by checking the commit status with  `pygit2.GIT_DELTA_RENAMED` ,`pygit2.GIT_DELTA_COPIED` , and `pygit2.GIT_DELTA_DELETED`  https://github.com/libgit2/pygit2/blob/d4e328de64a1e872f6a9f708443d8e7bb81c97a6/src/pygit2.c#L561 .

I guess the pain part is actually making changes to the DB according the commit status.

Reply:
Correct. I saw those constants, but the question is what all the scenarios are in which they need to be considered. For example, in [RustSec/advisory-db](https://github.com/RustSec/advisory-db) files are (sometimes?) added with a temporary name `RUSTSEC-0000-0000.toml`.

Reply:
Interesting case, could that be handled by passing some special config?

Reply:
What do you mean with special config? I don't know how it could be handled, that's why I added a `TODO` comment instead of writing the code for it. :)

Reply:
As this is a corner case, we could handle it later as you said. 
What I had in mind, is to pass in JSON data like `{'ignore':['RUSTSEC-0000-0000']}` in  `data_source_cfg` while registering the Importer model for rust.

As this `data_source_cfg` while be passed in to instantiate the corresponding `data_source` in the Importer model's `make_data_source` method, thus this  `data_source` has access to its `data_source_cfg`.

Now the data source for rust will most probably inherit from `GitDataSource` , we could then override/implement ` added_advisories` and `updated_advisories` such that it doesn't include vulnerabilities which have ids present in it's `data_source_cfg` 's  'ignore' list .

I hope I make sense :)

Reply:
Ah, now I understood what you mean with special config. Yes, that should work. I just finished the Rust importer based on `GitDataSource` and just filter the files in the code there (feedback welcome on that as well of course). But it could certainly be generalized like that if we need it in many importers.